File: BMWCenter/Core/Care/BaselineLearner.swift
import Foundation
import SwiftData

/// Online Welford mean/variance + 128-bin histogram for P50/P95.
@MainActor
final class BaselineLearner {
    static let histogramBins = 128
    static let halfLifeDays: Double = 60

    private let modelContext: ModelContext
    private var cache: [String: BaselineMetric] = [:]
    private var histograms: [String: [Int]] = [:]

    init(modelContext: ModelContext) {
        self.modelContext = modelContext
        warmCache()
    }

    func warmCache() {
        let rows = (try? modelContext.fetch(FetchDescriptor<BaselineMetric>())) ?? []
        cache.removeAll()
        histograms.removeAll()
        for row in rows {
            let k = compositeKey(row.key, row.bucketKey)
            cache[k] = row
            if let data = row.histogramData,
               let bins = try? JSONDecoder().decode([Int].self, from: data),
               bins.count == Self.histogramBins {
                histograms[k] = bins
            } else {
                histograms[k] = Array(repeating: 0, count: Self.histogramBins)
            }
        }
    }

    func observe(
        key: String,
        value: Double,
        bucketKey: String = "",
        minSamples: Int = 60,
        range: ClosedRange<Double> = -50...200,
        now: Date = .now
    ) {
        let ck = compositeKey(key, bucketKey)
        let metric = cache[ck] ?? {
            let m = BaselineMetric(key: key, bucketKey: bucketKey)
            modelContext.insert(m)
            cache[ck] = m
            histograms[ck] = Array(repeating: 0, count: Self.histogramBins)
            return m
        }()

        applyForgetting(metric: metric, now: now)

        metric.count += 1
        let delta = value - metric.mean
        metric.mean += delta / Double(metric.count)
        let delta2 = value - metric.mean
        metric.m2 += delta * delta2
        metric.lastUpdated = now

        var bins = histograms[ck] ?? Array(repeating: 0, count: Self.histogramBins)
        let clamped = min(max(value, range.lowerBound), range.upperBound)
        let span = range.upperBound - range.lowerBound
        let idx = min(Self.histogramBins - 1, max(0, Int((clamped - range.lowerBound) / span * Double(Self.histogramBins))))
        bins[idx] += 1
        histograms[ck] = bins
        metric.histogramData = try? JSONEncoder().encode(bins)

        let (p50, p95) = percentiles(bins: bins, range: range)
        metric.p50 = p50
        metric.p95 = p95
        metric.isMature = metric.count >= minSamples
        try? modelContext.save()
    }

    func snapshot(key: String, bucketKey: String = "") -> BaselineMetric? {
        cache[compositeKey(key, bucketKey)]
    }

    func isMature(key: String, bucketKey: String = "", minSamples: Int = 60) -> Bool {
        guard let m = snapshot(key: key, bucketKey: bucketKey) else { return false }
        return m.count >= minSamples
    }

    func reset(keyPrefix: String? = nil) {
        let rows = (try? modelContext.fetch(FetchDescriptor<BaselineMetric>())) ?? []
        for row in rows {
            if let keyPrefix, !row.key.hasPrefix(keyPrefix) { continue }
            modelContext.delete(row)
            let ck = compositeKey(row.key, row.bucketKey)
            cache.removeValue(forKey: ck)
            histograms.removeValue(forKey: ck)
        }
        try? modelContext.save()
    }

    /// Pure helper for tests — update in-memory Welford + histogram.
    nonisolated static func updateOnline(
        count: inout Int,
        mean: inout Double,
        m2: inout Double,
        bins: inout [Int],
        value: Double,
        range: ClosedRange<Double>
    ) {
        count += 1
        let delta = value - mean
        mean += delta / Double(count)
        let delta2 = value - mean
        m2 += delta * delta2
        if bins.count != histogramBins {
            bins = Array(repeating: 0, count: histogramBins)
        }
        let clamped = min(max(value, range.lowerBound), range.upperBound)
        let span = range.upperBound - range.lowerBound
        let idx = min(histogramBins - 1, max(0, Int((clamped - range.lowerBound) / span * Double(histogramBins))))
        bins[idx] += 1
    }

    nonisolated static func percentiles(bins: [Int], range: ClosedRange<Double>) -> (p50: Double, p95: Double) {
        let total = bins.reduce(0, +)
        guard total > 0 else { return (0, 0) }
        let span = range.upperBound - range.lowerBound
        func value(atQuantile q: Double) -> Double {
            let target = Double(total) * q
            var acc = 0
            for (i, c) in bins.enumerated() {
                acc += c
                if Double(acc) >= target {
                    return range.lowerBound + (Double(i) + 0.5) / Double(histogramBins) * span
                }
            }
            return range.upperBound
        }
        return (value(atQuantile: 0.50), value(atQuantile: 0.95))
    }

    private func percentiles(bins: [Int], range: ClosedRange<Double>) -> (Double, Double) {
        Self.percentiles(bins: bins, range: range)
    }

    private func applyForgetting(metric: BaselineMetric, now: Date) {
        let ageDays = now.timeIntervalSince(metric.lastUpdated) / 86_400
        guard ageDays > 1, metric.count > 0 else { return }
        // Exponential half-life: weight decays toward 0.5 after ~60 days of inactivity on the series.
        let weight = pow(0.5, ageDays / Self.halfLifeDays)
        if weight < 0.99 {
            let newCount = max(1, Int(Double(metric.count) * weight))
            metric.m2 *= weight
            metric.count = newCount
        }
    }

    private func compositeKey(_ key: String, _ bucket: String) -> String {
        bucket.isEmpty ? key : "\(key)|\(bucket)"
    }
}

File: mobile/src/core/care/baselineLearner.ts
import { eq } from "drizzle-orm";
import { db } from "../storage/db";
import { baselineMetrics } from "../storage/schema";

const HISTOGRAM_BINS = 128;
const HALF_LIFE_DAYS = 60;

interface CachedMetric {
  id: number;
  key: string;
  bucketKey: string;
  count: number;
  mean: number;
  m2: number;
  p50: number;
  p95: number;
  lastUpdated: number;
  isMature: boolean;
  bins: number[];
}

function compositeKey(key: string, bucket: string): string {
  return bucket ? `${key}|${bucket}` : key;
}

function percentiles(bins: number[], range: [number, number]): { p50: number; p95: number } {
  const total = bins.reduce((a, b) => a + b, 0);
  if (total === 0) return { p50: 0, p95: 0 };
  const [lo, hi] = range;
  const span = hi - lo;
  const valueAt = (q: number) => {
    const target = total * q;
    let acc = 0;
    for (let i = 0; i < bins.length; i++) {
      acc += bins[i]!;
      if (acc >= target) return lo + ((i + 0.5) / HISTOGRAM_BINS) * span;
    }
    return hi;
  };
  return { p50: valueAt(0.5), p95: valueAt(0.95) };
}

/** Online Welford mean/variance + 128-bin histogram for P50/P95, persisted to SQLite. */
class BaselineLearner {
  private cache = new Map<string, CachedMetric>();
  private loaded = false;

  private async ensureLoaded() {
    if (this.loaded) return;
    this.loaded = true;
    const rows = await db.select().from(baselineMetrics);
    for (const row of rows) {
      const ck = compositeKey(row.key, row.bucketKey);
      let bins: number[] = [];
      try {
        bins = row.histogramJSON ? JSON.parse(row.histogramJSON) : [];
      } catch {
        bins = [];
      }
      if (bins.length !== HISTOGRAM_BINS) bins = new Array(HISTOGRAM_BINS).fill(0);
      this.cache.set(ck, {
        id: row.id,
        key: row.key,
        bucketKey: row.bucketKey,
        count: row.count,
        mean: row.mean,
        m2: row.m2,
        p50: row.p50,
        p95: row.p95,
        lastUpdated: row.lastUpdated,
        isMature: row.isMature,
        bins,
      });
    }
  }

  async observe(
    key: string,
    value: number,
    bucketKey = "",
    minSamples = 60,
    range: [number, number] = [-50, 200],
    now = Date.now()
  ): Promise<void> {
    await this.ensureLoaded();
    const ck = compositeKey(key, bucketKey);
    let metric = this.cache.get(ck);
    if (!metric) {
      metric = { id: -1, key, bucketKey, count: 0, mean: 0, m2: 0, p50: 0, p95: 0, lastUpdated: now, isMature: false, bins: new Array(HISTOGRAM_BINS).fill(0) };
    }

    const ageDays = (now - metric.lastUpdated) / 86_400_000;
    if (ageDays > 1 && metric.count > 0) {
      const weight = Math.pow(0.5, ageDays / HALF_LIFE_DAYS);
      if (weight < 0.99) {
        metric.m2 *= weight;
        metric.count = Math.max(1, Math.round(metric.count * weight));
      }
    }

    metric.count += 1;
    const delta = value - metric.mean;
    metric.mean += delta / metric.count;
    const delta2 = value - metric.mean;
    metric.m2 += delta * delta2;
    metric.lastUpdated = now;

    const [lo, hi] = range;
    const span = hi - lo;
    const clamped = Math.min(Math.max(value, lo), hi);
    const idx = Math.min(HISTOGRAM_BINS - 1, Math.max(0, Math.floor(((clamped - lo) / span) * HISTOGRAM_BINS)));
    metric.bins[idx] = (metric.bins[idx] ?? 0) + 1;

    const { p50, p95 } = percentiles(metric.bins, range);
    metric.p50 = p50;
    metric.p95 = p95;
    metric.isMature = metric.count >= minSamples;
    this.cache.set(ck, metric);

    const histogramJSON = JSON.stringify(metric.bins);
    if (metric.id === -1) {
      const inserted = await db
        .insert(baselineMetrics)
        .values({
          key,
          bucketKey,
          count: metric.count,
          mean: metric.mean,
          m2: metric.m2,
          p50,
          p95,
          lastUpdated: now,
          isMature: metric.isMature,
          histogramJSON,
        })
        .returning({ id: baselineMetrics.id });
      metric.id = inserted[0]?.id ?? -1;
    } else {
      await db
        .update(baselineMetrics)
        .set({ count: metric.count, mean: metric.mean, m2: metric.m2, p50, p95, lastUpdated: now, isMature: metric.isMature, histogramJSON })
        .where(eq(baselineMetrics.id, metric.id));
    }
  }

  snapshot(key: string, bucketKey = ""): CachedMetric | undefined {
    return this.cache.get(compositeKey(key, bucketKey));
  }

  isMature(key: string, bucketKey = "", minSamples = 60): boolean {
    const m = this.snapshot(key, bucketKey);
    return m != null && m.count >= minSamples;
  }

  async reset(keyPrefix?: string): Promise<void> {
    await this.ensureLoaded();
    const toDelete = [...this.cache.values()].filter((m) => !keyPrefix || m.key.startsWith(keyPrefix));
    for (const m of toDelete) {
      if (m.id !== -1) await db.delete(baselineMetrics).where(eq(baselineMetrics.id, m.id));
      this.cache.delete(compositeKey(m.key, m.bucketKey));
    }
  }
}

export const baselineLearner = new BaselineLearner();
